### PR TITLE
Allow repeatable interceptor bindings

### LINKED ISSUE #13332: Repeated @InterceptorBinding are ignored by arc-processor
I struggled for a while today with a repeatable interceptor binding that was ignored by arc-processor until I found in debug mode why it is ignored.

Let's assume the interceptor binding looks like this:
```
@InterceptorBinding
@Retention(RetentionPolicy.RUNTIME)
@Target({ ElementType.TYPE, ElementType.METHOD })
@Repeatable(Bar.class)
public @interface Foo {

    @Nonbinding
    String someValue() default "";

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.METHOD)
    @interface Bar {
        Foo[] value();
    }
}
```

And is used that way:
```
@ApplicationScoped
public class EmptyService {
    @Foo(someValue = "one")
    @Foo(someValue = "two")
    public void doSomething() {
    }
}
```

Then, in the following code the repeated interceptor bindings will be ignored because `beanDeployment` doesn't know any interceptor binding named `Foo$Bar`:
https://github.com/quarkusio/quarkus/blob/dda2f83a1c0b1ca65884e4a0b65fc80070e57418/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java#L158-L160

Here's the value of `methodAnnotations` during the code execution:
```
[@Bar(value = [@Foo(someValue = "1"),@Foo(someValue = "2")])]
```

@mkouba Am I wrong trying to use repeatable interceptor bindings or is this a bug? It currently works when the bindings are added using an `AnnotationsTransformer` but I'd like it to work without the transformer as well.

### PR DESCRIPTION
**Pull request by Ladicek**
Fixes #13332.

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
Comment (manovotn) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: Isn't having multiple bindings of the same type (and all on class level) an error according to specification?

Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: That's how I read the spec, indeed -- but it doesn't look like we enforce it. (At least not yet.)

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: As per Zulip conversation - this should be validated and the spec requires this because you can, for instance, declare the same binding with member on a class and on a stereotype that this class uses leading to a conflict.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java
Comment (mkouba) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:45]: Shouldn't the expected value be `11` because `ClassInterceptedBean#foo()` declares `@MyBinding("foo")` and `@MyBinding("bar")` is inherited from the superclass?

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: +1 for validation and potentially a build failure, i.e. repeatable bindings are only allowed on methods which should be documented here https://quarkus.io/guides/cdi-reference#nonstandard_features

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java
Comment (Ladicek) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:45]: Exactly the kind of feedback I was looking for!

My understanding is that if you declare an interceptor binding on a method, interceptor bindings of the same type coming from the class and all superclasses (and even stereotypes I think) are ignored.

This is per https://jakarta.ee/specifications/interceptors/2.0/interceptors-spec-2.0.html#binding_an_interceptor_to_a_component:

> An interceptor binding declared on a method or constructor replaces an interceptor binding of the same type declared at class level or inherited from a superclass [7].

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java
Comment (mkouba) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java:62]: Hm, I don't think this should be supported. A stereotype with interceptor bindings is inherently a class-level-only concept and so I would expect the same limitations as with bindings declared on a class...

Comment (Ladicek) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java:62]: The stereotype is class-only.

The test also has methods with the same interceptor binding, verifying that if a method has an interceptor binding, other interceptor bindings (from the class, from superclasses, from stereotype) don't apply, but that's a valid case IMHO?

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java
Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:45]: You're right but it's not very intuitive :-(. This interceptors rule is another one that was created before repeatable annotations were introduced...

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java
Comment (Ladicek) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java:62]: (Ah and this test is `@Disabled`, because we don't do stereotype inheritance yet. But it's still valid to discuss I believe.)

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java
Comment (Ladicek) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:45]: Yea I totally agree, the interceptors spec needs an update wrt repeatable annotations and default interface methods.

Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:45]: Yet another reason to ditch the support of the class-level repeatable bindings...

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java
Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java:62]: My point is that a stereotype with multiple bindings of the same type should probably result in a definition exception...

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: OK, only supporting repeatable interceptor bindings on methods, that's an interesting proposition. @manovotn does perhaps Weld do the same? That would explain why there's no test for repeatable interceptor bindings on classes.

Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: > does perhaps Weld do the same? That would explain why there's no test for repeatable interceptor bindings on classes.

We should find out ;-)

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: Did you just volunteer to try that you? :-P

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: A quick remake of the existing Weld test shows me that interception happens even if I put repeatable annotation on a class-level of the bean.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java
Comment (manovotn) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:45]: Agree. The test is "correct" per spec but it doesn't follow common sense.
Since the specification is blurred in this regard (e.g. repeatable ann), we may IMO choose to go either way.
FTR I did another quick test in Weld and Weld sides with Ladislav on this one as it also chooses not to inherit repeating annotation if declared on the current class.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: Good to know. Doesn't necessarily mean we have to do the same, but I think the current behavior is actually pretty reasonable. (Once you make peace with the fact that whenever you put an interceptor binding on a method, all "inherited" interceptor bindings of the same type are ignored :-) )

Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:453]: > Doesn't necessarily mean we have to do the same

No, but for most people familiar with CDI, it would then behave in an "expected manner". E.g. if they used it before, they likely did so on Weld and aligning it would provide consistency. So I'd be +1 for that.

**Comment by Ladicek** [comment]
I have removed the support for class-level (and stereotype-level) repeatable interceptor bindings, together with the corresponding tests, and added a short section into the documentation.

I didn't add validatation for the "multiple class-level interceptor bindings of the same type" situation. We didn't do that before, so it's not a regression :-) If you think this must go in, let me know and I'll take a look. (Next year, probably :-) )

**Comment by mkouba** [comment]
> I have removed the support for class-level (and stereotype-level) repeatable interceptor bindings, together with the corresponding tests, and added a short section into the documentation.

+1

> I didn't add validatation for the "multiple class-level interceptor bindings of the same type" situation. We didn't do that before, so it's not a regression :-) If you think this must go in, let me know and I'll take a look. (Next year, probably :-) )

This should be addressed in a separate issue/PR anyway.

**Comment by Ladicek** [comment]
CI passed in my fork, so marked ready for review.

File: docs/src/main/asciidoc/cdi-reference.adoc
Comment (mkouba) [inline docs/src/main/asciidoc/cdi-reference.adoc:757]: Maybe we should clarify that repeatable interceptor bindings _declared on interceptor classes_ are supported?

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:757]: So this can get misleading pretty quickly IMHO. You always need to put an interceptor binding on two places:

- on the interceptor class;
- on the place that should be intercepted.

I don't know if there's a good name for the 2nd item, because that's what we would need here.

I don't think we should say anything like "Repeatable interceptor bindings declared on classes are not supported, unless the class is an interceptor class", because that conflates the two concepts.

Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc:757]: >     * on the interceptor class;

-> **Declaring Interceptor Bindings of an Interceptor**

>     * on the place that should be intercepted.

-> **Binding an Interceptor to a Component**

This is the spec wording. So I think that we support "Repeatable interceptor bindings declared on methods of component classes and interceptor classes"...

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:757]: Gotcha, I'll try to figure something out.

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:757]: WDYT about this?

```adoc
=== Repeatable interceptor bindings

Quarkus has a limited support for `@Repeatable` interceptor binding annotations.

When binding an interceptor to a component, you can declare multiple `@Repeatable` annotations on methods.
Repeatable interceptor bindings declared on classes and stereotypes are not supported, because there are some open questions around interactions with the Interceptors specification.
This might be added in the future.

As an example, suppose we have an interceptor that stores method's return values into a cache.
The corresponding interceptor binding would be called `@Cached` and would be declared as `@Repeatable`.
If we wanted to store the return value into two caches at the same time, we would add `@Cached` twice:

[source,java]
----
@ApplicationScoped
class AllEncompassingService {
  @Cached(cacheName = "foo")
  @Cached(cacheName = "bar")
  BigData heavyComputation() {
    // ...
  }
}
----

This is how interceptors are used.
What about creating an interceptor?

When declaring interceptor bindings of an interceptor, you can add multiple `@Repeatable` annotations to the interceptor class as usual.
This is useless when the annotation members are `@Nonbinding`, as would be the case for the `@Cached` annotation, but is important otherwise.

For example, suppose we have an interceptor that can automatically log method invocations to certain targets.
The interceptor binding annotation `@Logged` would have a member called `target`, which specifies where to store the log.
Our implementation could be restricted to console logging and file logging:

[source,java]
----
@Interceptor
@Logged(target = "console")
@Logged(target = "file")
class NaiveLoggingInterceptor {
  // ...
}
----

Other interceptors could be provided to log method invocations to different targets.
```

**Review by gwenneg — COMMENTED** [review]
This looks good, thanks for taking care of it! I added a few comments.

File: docs/src/main/asciidoc/cdi-reference.adoc
Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:746]: AsciiDoc actually understands `// <1>` and removes the `//`. Much better, since my IDE knows it's a Java code in that `.adoc` file and gives me syntax highlight and errors and everything -- I don't know why noone else does this! :-)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:425]: Ah, I'll take a look at that, I copied the existing method without much thought :-)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:156]: Thanks, I'll do that!

**Comment by Ladicek** [comment]
Actually, I really need to move to some other work, so I'll leave this to Monday, if everyone's fine with that :-)

File: docs/src/main/asciidoc/cdi-reference.adoc
Comment (gwenneg) [reply docs/src/main/asciidoc/cdi-reference.adoc:746]: That's good to know, thanks for the info!

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:156]: I'm looking at it again and I think this is a pretty good name. It takes annotation types and returns the containing annotation types (for each annotation in the original set that is `@Repeatable`). The term "containing annotation type" is what `@Repeatable` javadoc uses, so "container annotation" is fairly close.

I could rename to `findContainingAnnotationTypesForRepeatableAnnotationTypes` or something similarly long, but I don't think that's necessary. I've just renamed the parameters and local variables, as the method was originally written only for qualifiers, but it's now used in another setting as well.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:425]: Fixed this.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (Ladicek) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:428]: It's either "to be inspected", or "to be extracted from". In my (arguably poor) understanding of English, "annotation instance to be extraced" means it's already a result of the extraction process, which it's not; it's an input to the extraction process.

**Comment by Ladicek** [comment]
Rebased and pushed a commit that addresses the documentation feedback and @gwenneg's comments (most of them, anyway).

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (gwenneg) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:156]: Now that I checked the `Repeatable` javadoc, I agree with you. The current name is good enough and I don't think the extra long and explicit version would improve the code readability. Let's leave it that way! :)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
Comment (gwenneg) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: You're right and on second thoughts I don't think there's a problem with this line. I'm fine with the current version.

**Comment by Ladicek** [comment]
Rebased and resolved the conflict.

File: docs/src/main/asciidoc/cdi-reference.adoc
Comment (gwenneg) [reply docs/src/main/asciidoc/cdi-reference.adoc:755]: There's another reason why `@Cached` (even the imaginary version) might not be the best pick for that example: it doesn't really make sense to repeat an annotation aimed at retrieving data from a cache because having multiple caches on a single method would raise a lot of questions. That's probably why most (if not all) annotations-based caching API do not allow this.

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:755]: Yea I know it's stupid. The other example too. If you have a better one, I'm all ears :-)

File: docs/src/main/asciidoc/cdi-reference.adoc
Comment (gwenneg) [inline docs/src/main/asciidoc/cdi-reference.adoc:758]: Here's a proposal which is really close from the current Quarkus caching API.

```suggestion
As an example, suppose we have an interceptor that removes all entries from a cache.
The corresponding interceptor binding would be called `@CacheInvalidateAll` and would be declared as `@Repeatable`.
If we wanted to clear two caches at the same time, we would add `@CacheInvalidateAll` twice:

[source,java]
----
@ApplicationScoped
class AllEncompassingService {
  @CacheInvalidateAll(cacheName = "foo")
  @CacheInvalidateAll(cacheName = "bar")
  void clearBothCaches() {
    // ...
  }
```

When this PR is merged, `@CacheInvalidateAll` will even become an interceptor binding (it's a simple annotation right now).

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:758]: OK, I didn't really want to use _existing_ annotations, but this will do, thanks! :-)

**Comment by Ladicek** [comment]
Updated an example per @gwenneg's suggestion. I think this PR is ready now.

**Review by gwenneg — APPROVED** [review]
LGTM, thanks!

Let's wait for @mkouba or @manovotn approval before it is merged though :)

**Comment by mkouba** [comment]
I'm not quite sure we want to merge those disabled tests. Ideally, we should fix the problem in a separate PR ;-). @Ladicek WDYT?

**Comment by Ladicek** [comment]
Removed disabled tests. I kept them locally in case anyone is interested in the future :-)

**Review by mkouba — APPROVED** [review]

**Review by manovotn — APPROVED** [review]

**Comment by gwenneg** [comment]
Thanks everyone for that fix!

**Pull request merged**
